File: 121_best_time_to_buy_and_sell_stock/solution.ts
function maxProfit(prices: number[]): number {
  if (prices.length <= 1) return 0;

  let [l, r] = [0, 1];
  let maxP = 0;

  while (r < prices.length) {
    if (prices[l] < prices[r]) {
      const profit = prices[r] - prices[l];
      maxP = Math.max(maxP, profit);
    } else {
      l = r;
    }
    r++;
  }

  return maxP;
}

File: 121_best_time_to_buy_and_sell_stock/solution.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 1:
            return 0

        l, r = 0, 1
        maxPrice = 0

        while r < len(prices):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                maxPrice = max(maxPrice, profit)
            else:
                l = r
            r += 1

        return maxPrice
